### ISSUE TITLE
Disabling Sounds and Voicing in Preferences can result in the Audio button in the nav bar having no effect

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/joist#734]
This was discussed during a Voicing meeting today, and came up again during discussion in https://github.com/phetsims/joist/issues/724. If sub-features under the "Audio Features" are disabled in Preferences, it can result in cases where pressing the "Audio" button in the navigation bar will have no effect. Since both Voicing and Sounds are disabled, no audio will come out of the sim regardless of whether all audio is enabled/disabled. 

Maybe if both sounds and voicing are disabled the audio button in the navigation bar also becomes disabled? I am not sure what is best, but assigning to @emily-phet and @terracoda to consider a solution or if anything should be done.

**Comment by jessegreenberg** [comment]
Discussed today with @emily-phet and @terracoda. We thought of a few potential solutions
- When toggling the audio features back on in this state you could always re-enable sound.
- Maybe you HAVE to have at least one sound feature enabled when "Audio" is enabled.
- Add a third state to the NavigationBarSoundButton that indicates that all Audio features are disabled in preferences.

We don't like that for the first two options making one choice forces another. It is impossible for the user to be in the state they want.

We liked the third option the best. Instead of adding a new icon to indicate this, we tried making the NavigationBarSoundButton disabled:
![image](https://user-images.githubusercontent.com/6396244/125676632-b9f71d8f-796a-4c6f-8aa6-4e7b80e24286.png)

This seems like it covers the case well! We would like for it to behave like this. When Voicing and Sounds are turned off, the button should be `enabled: false`, so that it is not pickable and slightly transparent.

**Comment by jessegreenberg** [comment]
I will work on making this change to the NavigationBarSoundToggleButton.

**Comment by emily-phet** [comment]
@jbphet @arouinfar 

Hey folks - the solution we've come up with above (see screenshot) involves a visual display change to the audio on/off button in the nav bar. Can you take a look and give a thumbs up / thumbs down using the opaqueness of the audio on/off button to indicate "disabled"?

**Comment by jessegreenberg** [comment]
This patch would implement this change:
<details>

```patch
Index: js/NavigationBarSoundToggleButton.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/NavigationBarSoundToggleButton.js	(revision 71329d30a33349fa8b01d0bdbb8a33108cb10e7a)
+++ js/NavigationBarSoundToggleButton.js	(date 1626294246581)
@@ -8,13 +8,17 @@
  * @author Chris Klusendorf
  */
 
+import DerivedProperty from '../../axon/js/DerivedProperty.js';
+import Property from '../../axon/js/Property.js';
 import Vector2 from '../../dot/js/Vector2.js';
 import Shape from '../../kite/js/Shape.js';
 import merge from '../../phet-core/js/merge.js';
+import webSpeaker from '../../scenery/js/accessibility/voicing/webSpeaker.js';
 import Node from '../../scenery/js/nodes/Node.js';
 import Path from '../../scenery/js/nodes/Path.js';
 import Rectangle from '../../scenery/js/nodes/Rectangle.js';
 import ToggleNode from '../../sun/js/ToggleNode.js';
+import soundManager from '../../tambo/js/soundManager.js';
 import ActivationUtterance from '../../utterance-queue/js/ActivationUtterance.js';
 import joist from './joist.js';
 import JoistButton from './JoistButton.js';
@@ -135,7 +139,15 @@
     );
     soundOnNode.addChild( soundOnCurves );
 
-    const toggleNode = new ToggleNode( soundEnabledProperty,
+    // Property that controls the graphic for the button - show the "on" icon when all audio is enabled and at least one audio feature is enabled to make it clear that there will be no auditory output otherwise
+    const toggleNodeProperty = new DerivedProperty( [ soundEnabledProperty, soundManager.enabledProperty, webSpeaker.enabledProperty ],
+      ( audioEnabled, soundEnabled, voicingEnabled ) => {
+        return audioEnabled && ( soundEnabled || voicingEnabled );
+      }
+    );
+
+    const toggleNode = new ToggleNode( toggleNodeProperty,
       [
         { value: true, node: soundOnNode },
         { value: false, node: soundOffNode }
@@ -163,6 +175,12 @@
     soundEnabledProperty.lazyLink( pressedListener );
     this.setPDOMAttribute( 'aria-pressed', !soundEnabledProperty.get() );
 
+    // the button is disabled if both sound and voicing features are disabled - pressing this button
+    // will have no effect in this case so disabling makes this state clear
+    Property.multilink( [ soundManager.enabledProperty, webSpeaker.enabledProperty ], ( soundEnabled, speechEnabled ) => {
+      this.enabled = soundEnabled || speechEnabled;
+    } );
+
     // change the icon so that it is visible when the background changes from dark to light
     backgroundColorProperty.link( backgroundColor => {
       const baseColor = backgroundColor === 'black' ? 'white' : 'black';
Index: js/JoistButton.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/JoistButton.js	(revision 71329d30a33349fa8b01d0bdbb8a33108cb10e7a)
+++ js/JoistButton.js	(date 1626291395764)
@@ -12,6 +12,7 @@
 import FocusHighlightPath from '../../scenery/js/accessibility/FocusHighlightPath.js';
 import Voicing from '../../scenery/js/accessibility/voicing/Voicing.js';
 import Node from '../../scenery/js/nodes/Node.js';
+import SceneryConstants from '../../scenery/js/SceneryConstants.js';
 import ButtonInteractionState from '../../sun/js/buttons/ButtonInteractionState.js';
 import PushButtonInteractionStateProperty from '../../sun/js/buttons/PushButtonInteractionStateProperty.js';
 import PushButtonModel from '../../sun/js/buttons/PushButtonModel.js';
@@ -39,6 +40,7 @@
 
       // JoistButtons by default do not have a featured enabledProperty
       enabledPropertyOptions: { phetioFeatured: false },
+      disabledOpacity: SceneryConstants.DISABLED_OPACITY,
 
       // pdom
       tagName: 'button'
```
</details>

I can apply it once this is cleared by @jbphet / @arouinfar.

**Comment by arouinfar** [comment]
I think disabling the sound toggle button is a reasonable solution, and the disabled button looks good to me. In https://github.com/phetsims/sun/issues/578 we decided to make `SunConstants.DISABLED_PROPERTY` 0.45 after reviewing a variety of UI components/contexts. Looks like that's now `SceneryConstants.DISABLED_PROPERTY` which is what @jessegreenberg included in the patch.

**Comment by jessegreenberg** [comment]
Thanks @arouinfar - I went ahead with this change in the above commit. Final commit is a bit different from the patch in https://github.com/phetsims/joist/issues/734#issuecomment-880188179. To summarize:

- It takes advantage of the new audioManager and observes Properties on it to know when to disable the nav bar button and change its icon.
- I made a change to the initial value of the soundManager.enabledProperty to respect `?supportsSound`. It previously only used `?sound`.  But now soundManager.enabledProperty is initially true if `?supportsSound=true` and `?sound=enabled`, so that the NavigationBarAudioToggleButton has the correct behavior with both of these query params.

@jbphet can you please review https://github.com/phetsims/joist/issues/734#issuecomment-880170846 as well as the code change?

**Comment by jbphet** [comment]
I think the concept is good and I'm up for this change.  I just tried it out on John Travoltage and Balloons and Static Electricity, because the former has voice and sound and the latter just has sound, and the behavior seems reasonable.  However, sounds are not working in either sim.  I don't know if that is related to this issue or not, but I'm not comfortable closing it until sound is fully functional.

I'm going to assign this back to @jessegreenberg to investigate and will ping him on Slack.  I'll look at it myself later today if it is not resolved by then.

**Comment by jbphet** [comment]
Wait, never mind - I'd muted the tab.  The behavior seems fine.  Closing.

**Comment by jbphet** [comment]
Reopening.  I just came across a problem with the behavior of the nav bar icon and its interaction with query parameters.  The query parameter for `sound` has valid values of `[ 'enabled', 'disabled', 'muted' ]`.  I just tried using `muted` on greenhouse-effect, and the sound is indeed muted at the beginning, but the icon is grayed out and there is no way to unmute it.  This shouldn't be the behavior.  The sound should be _initially_ muted when this setting is used, but it should be possible to unmute it.

**Comment by jessegreenberg** [comment]
This is expected with the current UI because the navigation bar button doesn't control "Sounds" anymore, it controls "Audio". The new way to mute initially but still re-enable with the NavigationBarAudioToggleButton is to go through the `audio` query parameter, which supports the same values `[ 'enabled', 'disabled', 'muted' ]`.

With this design, there is no way to unmute sounds with `sound=muted` query param, unless the sim has a preferences dialog and the preferences dialog has a "Sounds" toggle button.

**Comment by emily-phet** [comment]
@jbphet - perhaps there is an issue with the current setup for the query parameters? The change from sound to all audio control may mean the sound query parameters should be deprecated as sound becomes a subset of all audio, and maybe needs a new structure to the query parameters.

**Comment by jbphet** [comment]
@jessegreenberg said:
 
> The new way to mute initially but still re-enable with the NavigationBarAudioToggleButton is to go through the audio query parameter

... and @emily-phet said:

> @jbphet - perhaps there is an issue with the current setup for the query parameters? The change from sound to all audio control may mean the sound query parameters should be deprecated as sound becomes a subset of all audio, and maybe needs a new structure to the query parameters.

I'm generally fine with this idea, but have few thoughts:
1. The `sound` query parameter is marked at public facing, and I don't know if it's cool to just change it.  I'm not even sure who to ask.  People might be using it out there on a regular basis and, if so, removing it could be a problem for them. 
 @jessegreenberg and @emily-phet - do you know how to find out who to ask about this?
2.  This query parameter is also used by aqua, our continuous test system, to prevent sims from making a bunch of noise during continuous testing.  If we deprecate or change this parameter, we'll need to make the appropriate adjustments there too.
3. I think we should make a clear decision - at least for now - as to whether we want a single query parameter that controls all audio and nothing specific for voicing and sound generation, or the audio parameter _and_ individual parameters for voicing and sound.  If we decide to have the latter, they should be designed and coded to work in a reasonable way when only one of the two audio features is present in a sim.

**Comment by emily-phet** [comment]
Some thoughts and questions for further discussion: 

- The original issue @jbphet brought up is only a problem *if the sim has no preferences menu*, correct? If so, the number of sims that have full sonification and no preferences menu will decrease over time. 

- If you use a query parameter that indicates you don't want sound, I'm not sure that I would be shocked to find I can't enable sound in the sim itself (which would be the case for sims with sound but no preferences menu). 

- Proposing perhaps the simplest solution - we no longer have a query parameter for sound that is "muted", but only "enabled/disabled". Is there a case (i.e., in testing, etc.) where the query parameter "muted" is critical?

**Comment by jessegreenberg** [comment]
> The sound query parameter is marked at public facing, and I don't know if it's cool to just change it....@jessegreenberg and @emily-phet - do you know how to find out who to ask about this?

I had the same thought and asked this on slack a little while ago. @arouinfar said that she would add the new `audio` query parameter to teacher tips. I was also told to add `audio` query parameter to https://docs.google.com/spreadsheets/d/1TddeTtD8zAayJzs02g0fL7dR1iWYczs1CMFdYd3SBNc/edit#gid=0. But I don't think anything was done about the navigation bar button and `sound` query param. I put a question on the status meeting agenda to ask the whole team what the policy is for changing public query params.

> The original issue @jbphet brought up is only a problem if the sim has no preferences menu, correct?

The issue is a problem if there is no preferences menu and if there is no voicing, because we decided to hide the "Sounds" toggle switch when only one audio feature is supported so the "Audio" and "Sounds" switch are not redundant in that case.

> Proposing perhaps the simplest solution - we no longer have a query parameter for sound that is "muted", but only "enabled/disabled"

This seems reasonable to me.

**Comment by jbphet** [comment]
> @jbphet what do you think?

It's fine with me as long as the designers are okay with it.

We should have a follow-up issue to update aqua, since it would probably be better to use `audio=disabled` for the automated testing.

**Comment by emily-phet** [comment]
@jbphet - Are you referring to @arouinfar and @amanda-phet? I can reach out to them about this if that's the case. I think you're referring to people who know about the public-facing nature of the query parameters, which would be the folks I name above, plus maybe Oliver and Kathy? 

Can you list the people or more specific attributes of the group of people you'd like input in on this? If needed, I could request a short design meeting slot or reach out to folks individually to get this sorted quickly.

**Comment by arouinfar** [comment]
I've never used the `sound` query parameter before, and I don't know if it is regularly used by partners/clients. 

When @jessegreenberg announced the `audio` query parameter, I made a note to add it to the Teacher Tips for Fourier: Making Waves. That sim has a sound, but no other audio. The current behavior is as I would expect -- `audio=disabled` permanently turns off sound and `audio=muted` mutes the sound by default but users can turn it on if they want. I see both of these options as useful. A teacher might want students to opt-in to sound (rather than opt-out). There may also be times when they just don't want to deal with sound at all, so disabling it is nice. 

There may be cases where `audio=disabled` leads to strange behavior in the sim. Waves Intro and Fourier both have checkboxes that control a sound, so these checkboxes would always be disabled. I still think it's fine for the query parameter to exist, but users would need to know exactly what passing `disabled` to the query parameter will do.

**Comment by emily-phet** [comment]
I'd like to pick a path forward for publication soon, but it sounds like we might need to be clear on how the possible solutions impact sims with different audio-related contents.
I'm going to request some design time on Thursday.

**Comment by jbphet** [comment]
@arouinfar, @terracoda, @jessegreenberg, and I discussed this in a meeting on 8/5/2021.  We decided to go with the minimalist approach, and have only one query parameter for now, the `audio` parameter.  Part of the justification for this decision is that if we tried to support one overarching query parameter like `audio` and then more feature-specific parameters like `sound`, it would be easy to create contradictory parameter strings, like `audio=disabled` and `sound=enabled`.

We will deprecate the `sound` parameter in `initialize-globals` now so that this change will be included in the upcoming releases of JT and BASE.  This doesn't change the functionality, and is just more of a documentation thing.

@jbphet will remove the `sound` query parameter and update the dependent code.  This will likely require coordination with @jonathanolson for updating the usage in aqua (continuous testing).  The phetmarks repo probably also uses it.

**Comment by jbphet** [comment]
I've removed the labels that were on this, since some work needs to be done before it's ready for review and because we decided that it is okay to publish with the current behavior since the `sound` query parameter is not known to be widely used outside of the development team.

**Comment by jbphet** [comment]
I've removed support for the `sound` query parameter, since it has been supplanted by the `audio` one.  I also removed support for `audioVolume`, which was related and is described in https://github.com/phetsims/vibe/issues/33.  There are no related problems that I can see on the continuous test (CT) server, and no one has reported any problems as a result after about 24 hours, so I'm going to close this.  I'll reopen or log a new issue if something related comes up.
